### ISSUE TITLE
Create a codelist KPI

### ISSUE THREAD

**Issue by gaubert** [wmo-im/wcmp#35]
For all value associated to a codelist, it should be verified that the value belongs to the referenced codelist.

**Comment by gaubert** [comment]
TODO:

* Insure that values from codelists belong to the referenced codelist. 
* Distinguish the two different implementations (ISO CodeLists and WMO CodeList).  Values in WMOCodelist should be described as an anchor.
(Link in the anchor should be verified).
* Provide guidance on how the thesaurus should be described in WMCP

**Comment by martaguti** [comment]
I Have left KPI H as a Keyword KPI that checks for well defined Thesaurus
I will create a KPI that checks for the presence of any of the WMOCodelist:
https://github.com/wmo-im/wmcp/tree/master/codelists
An ensures that values are permitted.

The KPI will have to distinguish between:

1. Extended codelist values from ISO (which are implemented via codeList and codeListValue attributes )

- CI_DateTypeCode
- MD_KeywordTypeCode

2. WMO specific codelist: (These will only be enforceable if we reference them with gmx:Anchors) 

- WMO_CategoryCode
- WMO_DataLicenseCode
- WMO_DistributionScopeCode
- WMO_GTSProductCategoryCode

**Comment by tomkralidis** [comment]
cc @martaguti 

Update:
- this will now be covered in [KPI M](https://github.com/wmo-im/wcmp/tree/KPI-M)
- codelists have been moved to https://github.com/wmo-im/wcmp-codelists

**Comment by martaguti** [comment]
@wmo-im/tt-wismd 

Updated KPI M for codelist validation 
https://github.com/wmo-im/wcmp/blob/KPI-M/KPI/M.adoc

We have an issue here,  ISO has moved all the 19139 schemas and codelists
All references to 
http://standards.iso.org/ittf/PubliclyAvailableStandards/ISO_19139_Schemas/resources/codelist/gmxCodelists.xml

in the WMO guidelines and WMOCodeLists.xml  are no longer valid

See github issue:
https://github.com/ISO-TC211/schemas/issues/24

Which is the authoritative list now  @amilan17 

1. https://standards.iso.org/iso/19139/resources/gmxCodelists.xml
2.https://schemas.isotc211.org/schemas/19139/resources/gmxCodelists.xml

**Comment by tomkralidis** [comment]
Thanks @martaguti .  See https://github.com/wmo-im/wcmp/issues/30#issuecomment-619393828 for another reference to the ISO schemas.

Perhaps we should clarify whether we restrict to allow only the official ISO and WMO codelists in a WCMP document, or whether **any** codelist value in a WCMP document is allowed, and, if specified, should be based on a valid codelist catalogue.

For example, here's a codelist catalogue we created for WOUDC: https://geo.woudc.org/codelists.xml, which we use for [WOUDC discovery metadata](https://geo.woudc.org/csw?service=CSW&version=2.0.2&request=GetRepositoryItem&id=urn%3Ax-wmo%3Amd%3Aint.wmo.wis%3A%3Ahttps%3A%2F%2Fgeo.woudc.org%2Fdef%2Fdata%2Fozone%2Ftotal-column-ozone%2Ftotalozone) (example snippet below)

```xml
      <gmd:descriptiveKeywords>
        <gmd:MD_Keywords>
          <gmd:keyword>
            <gco:CharacterString>dobson</gco:CharacterString>
          </gmd:keyword>
          <gmd:keyword>
            <gco:CharacterString>brewer</gco:CharacterString>
          </gmd:keyword>
          <gmd:keyword>
            <gco:CharacterString>vassey</gco:CharacterString>
          </gmd:keyword>
          <gmd:keyword>
            <gco:CharacterString>pion</gco:CharacterString>
          </gmd:keyword>
          <gmd:keyword>
            <gco:CharacterString>microtops</gco:CharacterString>
          </gmd:keyword>
          <gmd:keyword>
            <gco:CharacterString>spectral</gco:CharacterString>
          </gmd:keyword>
          <gmd:keyword>
            <gco:CharacterString>hoelper</gco:CharacterString>
          </gmd:keyword>
          <gmd:keyword>
            <gco:CharacterString>saoz</gco:CharacterString>
          </gmd:keyword>
          <gmd:keyword>
            <gco:CharacterString>filter</gco:CharacterString>
          </gmd:keyword>
          <gmd:type>
            <gmd:MD_KeywordTypeCode codeList="https://geo.woudc.org/codelists.xml#WOUDC_InstrumentCode" codeListValue="theme" codeSpace="ISOTC211/19115">theme</gmd:MD_KeywordTypeCode>
          </gmd:type>
        </gmd:MD_Keywords>
      </gmd:descriptiveKeywords>
```

**Comment by gaubert** [comment]
we have keywords from the Keywords from the WMO_CategoryCode code list and it is mandatory to pick one from that list. 
For the rest, it is up to the provider to add up keywords from a defined codelist. This is some additional information which can help the discovery or is in used by other catalogues. My opinion is that we should not restrict it but maybe indicate how the additional keywords should be added in the metadata (different instance of MD_Keywords with a well qualified thesaurus, ...).

**Comment by martaguti** [comment]
@tomkralidis @gaubert 

When I mention in the KPI to validate against authoritative lists , is not with the intention of restricting them.
It is to harmonise which one is the official version of the lists that we support (from ISO and WMOCodeLists)
As discussed in https://github.com/wmo-im/wcmp/issues/30#issuecomment-619393828  there are  several copies of the standards (for 19139) hosted at several locations. They may all be same (or not ) . Do we support them all ? 

In fact as mentioned previously the link that was referred to in the WMO guidelines and that is referred to in the WMOCodelist.xml is no longer valid, as ISO has moved the URLS 
http://standards.iso.org/ittf/PubliclyAvailableStandards/ISO_19139_Schemas/resources/codelist/gmxCodelists.xml

As for importing from other codelists e.g. to refer to additional keywords from different domains and catalogues, 
KPI H is showing examples when terms are referred to from other dictionaries other than MD_Keywords  (in our case WMO_CategoryCode) It does also lists other potential dictionaries relevant to the community.

When other codelists are used then we can only validate terms if they are referred to via an Anchor (with URL's well defined) 

A separate question:
How far we go for the validation of terms in CharacterStrings  (if we know that by the location in the WCMP element , it's meant to be a codelist with no URL ) ?

Marta

**Comment by tomkralidis** [comment]
Implemented in #86
